### DISCUSSION TITLE
add Coding Conventions

### DISCUSSION THREAD

**Discussion by moto6** [fosslight/fosslight#947] [Show and tell]
## abstract & summary

- There are many different styles of code in this open source, as many people have contributed with their personalities.
- Some of them may violate the coding conventions of the general Java community, so we want to create a system that enforces general Java community rules

<br><hr><br>

## description & How to

-  those are just my opinion, so feel free to discuss!  and.. i appreciate it if you could suggest your opinion
 
### point 1 ) why `Coding Conventions` ? 

- this discussion handle not only `check-style` but also `code formatter`
- The reason we bring these two subjects ( `check-style`  and `code formatter` ) together is that most of the points pointed out in the check-style are automatically corrected by the formatter, so repetitive and simple tasks can be delegated to the computer.

### point 2 ) about platform independency

- I am well aware of the concerns that this project will be dependent on a specific IDE, OS or platform 
- there are two most widely used IDEs in our tech (java&Spring) stack, `intelliJ IDEA` and `spring sts(=eclipse)`, and i am working on this problem in a way that supports all of them.

<br><hr><br>

## Action strategy 

- action strategy i think :  `divide` and `conquer`

### 1) Share & Notice only
- Share and notify checkstyle and codeformatter (some ide settings may be included) files. If you add these changes to git-mainline , you should be fine as-is (without affecting other parts of the code).

### 2) lean-startup
- From further contributions, it is induced to allow code that applies checkstyle & formatter (following convention).
- Apply step by step, starting with a very small part, the part that is changed and added.

### 3) How to apply to huge files
- For example, in the case of `LicenseController`, the file is close to 1000 lines, so when applying the code convention (checkstyle & formatter) to the file, git conflict will occur in all PRs that include modifications to the file. It's possible.
- Therefore, for all files with hundreds of lines, please observe the following precautions when applying this convention
  - Even if it is the same Java class file, the convention does not apply to the entire file.
  - Apply only the part you have changed and modified (the part you are interested in).

> in Korean,
### (한국어버전) 3) 거대한 소스파일에 컨벤션을 적용하는법
- 예를들어 `LicenseController` 클래스의 경우에는 대략 1000 라인정도 되는 클래스인데, 컨벤션 (checkstyle & formatter) 을 파일 전체에 적용해보리면, 해당 컨트롤러를 수정하는 모든 PR 들과 아직 수정중이라 PR 이 올라오지 않은 많은 오픈소스 기여자들의 작업물에서 git conflict 가 발생할 가능성이 있습니다.
- 그런 이유로, 파일/클래스 들은 (일시에 코드컨벤션을 적용하는게 아니라) 아래의 단계를 거쳐서 코드컨벤션을 적용하는게 어떨까 합니다
  - 수정한 부분과 동일한 파일(하나의 파일)내 위치해 있을지라도, 전체를 한번에 적용하지 않습니다.
  - 해당 PR 에서 수정/변경사항이 반영된 부분에 대해서만 코드컨벤션을 적용합니다. 

<br><hr><br>

### concluding

- i'm just one of developers, so If you have any directions or idea, and feel free to  leave a comment  and share your opinion

**Comment by brorica** [comment]
why don't you document the big differences between likes and dislikes of conventions that are often wrong or I think it's good to organize expressions that are often wrong in the pr process
